File: apps/worker/src/notifications-cost-sharing.test.ts
import { createDefaultAppSettings } from "@renewlet/shared/settings-defaults";
import type { ApiAppSettings } from "@renewlet/shared/schemas/settings";
import { apiSubscriptionSchema, type ApiSubscription } from "@renewlet/shared/schemas/subscriptions";
import { isValidDateOnly, type DateOnly } from "@renewlet/shared/runtime";
import { describe, expect, it, vi } from "vitest";
import { listNotificationScheduleCandidateSubscriptions } from "./db";
import { collectNotificationItemsForLocalDate } from "./notifications";
import type { Env } from "./types";

vi.mock("./smtp", () => ({
  notificationSmtpConfig: () => {
    throw new Error("SMTP should not be used by collection reminder tests");
  },
  sendSmtpEmail: async () => undefined,
}));

function settings(overrides: Partial<ApiAppSettings> = {}): ApiAppSettings {
  return {
    ...createDefaultAppSettings(),
    timezone: "UTC",
    notificationReminderDays: 5,
    ...overrides,
  };
}

function dateOnly(value: string): DateOnly {
  if (!isValidDateOnly(value)) throw new Error(`Invalid test date: ${value}`);
  return value as DateOnly;
}

function subscription(overrides: Partial<ApiSubscription> = {}): ApiSubscription {
  return apiSubscriptionSchema.parse({
    id: "sub_family",
    name: "Family Plan",
    price: "30",
    currency: "USD",
    billingCycle: "monthly",
    category: "productivity",
    status: "active",
    pinned: false,
    publicHidden: false,
    startDate: "2025-12-01",
    nextBillingDate: "2026-01-10",
    autoRenew: false,
    autoCalculateNextBillingDate: true,
    tags: [],
    reminderDays: 0,
    repeatReminderEnabled: false,
    repeatReminderInterval: "1h",
    repeatReminderWindow: "72h",
    extra: {},
    ...overrides,
  });
}

function d1All<T>(results: T[] = []): D1Result<T> {
  return { success: true, results, meta: {} as D1Meta } as D1Result<T>;
}

describe("Cloudflare cost sharing collection notifications", () => {
  it("emits collection reminders even when ordinary subscription reminders are disabled", () => {
    const items = collectNotificationItemsForLocalDate("2026-01-07", settings(), [
      subscription({
        reminderDays: -2,
        costSharing: {
          enabled: true,
          splitMode: "equal",
          collectionReminder: { enabled: true, reminderDays: 3 },
          members: [
            { id: "partner", name: "Partner", currency: "USD", joinedDate: dateOnly("2025-12-10") },
            { id: "child", name: "Child", currency: "USD", joinedDate: dateOnly("2025-12-10") },
          ],
        },
      }),
    ]);

    expect(items).toEqual([
      expect.objectContaining({
        type: "costSharing",
        reminderDays: 3,
        costSharing: { memberName: "Partner", amount: "10", currency: "USD" },
      }),
      expect.objectContaining({
        type: "costSharing",
        reminderDays: 3,
        costSharing: { memberName: "Child", amount: "10", currency: "USD" },
      }),
    ]);
  });

  it("uses member custom amount and currency without exchange-rate guessing", () => {
    const items = collectNotificationItemsForLocalDate("2026-01-05", settings(), [
      subscription({
        costSharing: {
          enabled: true,
          splitMode: "custom",
          collectionReminder: { enabled: true, reminderDays: -1 },
          members: [
            { id: "partner", name: "Partner", customAmount: "7.5", currency: "EUR", joinedDate: dateOnly("2025-12-10") },
          ],
        },
      }),
    ]);

    expect(items).toEqual([
      expect.objectContaining({
        type: "costSharing",
        reminderDays: 5,
        costSharing: { memberName: "Partner", amount: "7.5", currency: "EUR" },
      }),
    ]);
  });

  it("emits sharing seat renewal reminders with the seat currency and term amount", () => {
    const familySubscription = subscription({ reminderDays: -2 }) as ApiSubscription & {
      sharingSeatReminders: Array<{
        memberName: string;
        monthlyPrice: "15";
        currency: string;
        billingMonths: number;
        expiresAt: string;
      }>;
    };
    familySubscription.sharingSeatReminders = [{
      memberName: "Quarterly friend",
      monthlyPrice: "15",
      currency: "EUR",
      billingMonths: 3,
      expiresAt: "2026-01-12",
    }];

    const items = collectNotificationItemsForLocalDate("2026-01-07", settings(), [familySubscription]);

    expect(items).toEqual([
      expect.objectContaining({
        type: "costSharing",
        targetDate: "2026-01-12",
        reminderDays: 5,
        costSharing: { memberName: "Quarterly friend", amount: "45", currency: "EUR" },
      }),
    ]);
  });

  it("skips one-time buyouts but keeps fixed-term one-time collection reminders", () => {
    const items = collectNotificationItemsForLocalDate("2026-01-07", settings(), [
      subscription({
        id: "sub_buyout",
        billingCycle: "one-time",
        startDate: "2025-12-10",
        nextBillingDate: "2025-12-10",
        autoCalculateNextBillingDate: false,
        costSharing: {
          enabled: true,
          splitMode: "equal",
          collectionReminder: { enabled: false, reminderDays: 3 },
          members: [{ id: "partner", name: "Partner", currency: "USD", joinedDate: dateOnly("2025-12-10") }],
        },
      }),
      subscription({
        id: "sub_fixed_term",
        billingCycle: "one-time",
        oneTimeTermCount: 1,
        oneTimeTermUnit: "month",
        startDate: "2025-12-10",
        nextBillingDate: "2026-01-10",
        autoCalculateNextBillingDate: false,
        costSharing: {
          enabled: true,
          splitMode: "equal",
          collectionReminder: { enabled: true, reminderDays: 3 },
          members: [{ id: "partner", name: "Partner", currency: "USD", joinedDate: dateOnly("2025-12-10") }],
        },
      }),
    ]);

    expect(items).toEqual([
      expect.objectContaining({
        subscriptionId: "sub_fixed_term",
        type: "costSharing",
        costSharing: { memberName: "Partner", amount: "15", currency: "USD" },
      }),
    ]);
  });

  it("uses indexed D1 mirror columns for scheduled collection candidates", async () => {
    const queries: Array<{ sql: string; params: unknown[] }> = [];
    const env = {
      DB: {
        prepare: (sql: string) => ({
          bind: (...params: unknown[]) => {
            queries.push({ sql, params });
            return {
              all: async <T>() => d1All<T>([]),
            };
          },
        }),
      } as unknown as D1Database,
      ASSETS: {} as Fetcher,
      ASSETS_BUCKET: {} as R2Bucket,
    } satisfies Env;

    await listNotificationScheduleCandidateSubscriptions(env, "usr_due", {
      scheduledLocalDate: "2026-01-07",
      includeExpired: true,
      showExpired: true,
    });

    const query = queries[0];
    expect(query?.sql).toContain("UNION");
    expect(query?.sql).toContain("cost_sharing_collection_reminder_enabled = 1");
    expect(query?.sql).toContain("cost_sharing_next_collection_reminder_date <= ?");
    expect(query?.sql).toContain("JOIN sharing_seats seat");
    expect(query?.sql).toContain("seat.expires_at >= ? AND seat.expires_at <= ?");
    expect(query?.sql).not.toMatch(/json_extract|json_valid|\$\.collectionReminder/);
    expect(query?.params).toEqual(expect.arrayContaining(["usr_due", -2, "2026-01-07"]));
  });
});

File: apps/docker-server/cmd/renewlet/notifications_cost_sharing_test.go
package main

import (
	"errors"
	"strings"
	"testing"
	"time"

	"github.com/pocketbase/pocketbase/core"
	"github.com/pocketbase/pocketbase/tools/types"
)

func TestBuildDueNotificationCreatesCostSharingCollectionReminders(t *testing.T) {
	settings := defaultAppSettings()
	settings.LocalePreference = string(preferenceZhCN)
	settings.Timezone = "UTC"
	settings.NotificationReminderDays = 3
	reminderDays := inheritReminderDays

	message := buildDueNotificationForLocalDate("2026-05-14", time.Date(2026, 5, 14, 8, 0, 0, 0, time.UTC), settings, []notificationSubscription{
		{
			ID:              "family",
			Name:            "Family Plan",
			Price:           "30",
			Currency:        "USD",
			Status:          "active",
			BillingCycle:    "monthly",
			NextBillingDate: "2026-05-17",
			ReminderDays:    disabledReminderDays,
			CostSharing: costSharingPayload{
				Enabled:   true,
				SplitMode: "equal",
				CollectionReminder: &costSharingCollectionReminder{
					Enabled:      true,
					ReminderDays: &reminderDays,
				},
				Members: []costSharingMember{
					{ID: "partner", Name: "Partner", JoinedDate: "2026-04-17", Currency: "USD"},
					{ID: "friend", Name: "Friend", JoinedDate: "2026-04-17", Currency: "USD"},
				},
			},
		},
	}, true, accountContentLocale(settings))

	if !message.HasPayload || len(message.Items) != 2 {
		t.Fatalf("expected two collection reminders, got %#v", message.Items)
	}
	for _, item := range message.Items {
		if item.Type != "costSharing" || item.CostSharing == nil || item.CostSharing.Amount != "10" || item.CostSharing.Currency != "USD" {
			t.Fatalf("unexpected cost sharing item: %#v", item)
		}
	}
	if !strings.Contains(message.Content, "家庭共享收款") || strings.Contains(message.Content, "即将续费：") {
		t.Fatalf("expected only collection copy in content, got %q", message.Content)
	}
}

func TestBuildDueNotificationUsesCustomCostSharingCollectionCurrency(t *testing.T) {
	settings := defaultAppSettings()
	settings.Timezone = "UTC"
	reminderDays := 1
	customAmount := "42"

	message := buildDueNotificationForLocalDate("2026-05-16", time.Date(2026, 5, 16, 8, 0, 0, 0, time.UTC), settings, []notificationSubscription{
		{
			ID:              "family-custom",
			Name:            "Family Custom",
			Price:           "30",
			Currency:        "USD",
			Status:          "active",
			BillingCycle:    "monthly",
			NextBillingDate: "2026-05-17",
			ReminderDays:    disabledReminderDays,
			CostSharing: costSharingPayload{
				Enabled:   true,
				SplitMode: "custom",
				CollectionReminder: &costSharingCollectionReminder{
					Enabled:      true,
					ReminderDays: &reminderDays,
				},
				Members: []costSharingMember{
					{ID: "partner", Name: "Partner", JoinedDate: "2026-04-17", Currency: "CNY", CustomAmount: &customAmount},
				},
			},
		},
	}, true, accountContentLocale(settings))

	if !message.HasPayload || len(message.Items) != 1 {
		t.Fatalf("expected one custom collection reminder, got %#v", message.Items)
	}
	payload := message.Items[0].CostSharing
	if payload == nil || payload.MemberName != "Partner" || payload.Amount != "42" || payload.Currency != "CNY" {
		t.Fatalf("expected custom member amount/currency, got %#v", payload)
	}
}

func TestBuildDueNotificationCreatesSharingSeatRenewalReminder(t *testing.T) {
	settings := defaultAppSettings()
	settings.Timezone = "UTC"
	settings.NotificationReminderDays = 5

	message := buildDueNotificationForLocalDate("2026-09-25", time.Date(2026, 9, 25, 8, 0, 0, 0, time.UTC), settings, []notificationSubscription{
		{
			ID:              "netflix-1",
			Name:            "Netflix #1",
			Price:           "45",
			Currency:        "CNY",
			Status:          "active",
			BillingCycle:    "monthly",
			NextBillingDate: "2026-10-01",
			ReminderDays:    disabledReminderDays,
			SharingSeats: []notificationSharingSeat{
				{MemberName: "Quarterly friend", MonthlyPrice: "15", Currency: "USD", BillingMonths: 3, ExpiresAt: "2026-09-30"},
			},
		},
	}, true, accountContentLocale(settings))

	if !message.HasPayload || len(message.Items) != 1 {
		t.Fatalf("expected one sharing seat reminder, got %#v", message.Items)
	}
	item := message.Items[0]
	if item.Type != "costSharing" || item.TargetDate != "2026-09-30" || item.CostSharing == nil || item.CostSharing.MemberName != "Quarterly friend" || item.CostSharing.Amount != "45" || item.CostSharing.Currency != "USD" {
		t.Fatalf("unexpected sharing seat reminder: %#v", item)
	}
}

func TestBuildDueNotificationSkipsOneTimeBuyoutCostSharingCollection(t *testing.T) {
	settings := defaultAppSettings()
	settings.Timezone = "UTC"
	reminderDays := 0

	message := buildDueNotificationForLocalDate("2026-05-14", time.Date(2026, 5, 14, 8, 0, 0, 0, time.UTC), settings, []notificationSubscription{
		{
			ID:              "lifetime-family",
			Name:            "Lifetime Family",
			Price:           "30",
			Currency:        "USD",
			Status:          "active",
			BillingCycle:    "one-time",
			NextBillingDate: "2026-05-14",
			ReminderDays:    disabledReminderDays,
			CostSharing: costSharingPayload{
				Enabled:   true,
				SplitMode: "equal",
				CollectionReminder: &costSharingCollectionReminder{
					Enabled:      true,
					ReminderDays: &reminderDays,
				},
				Members: []costSharingMember{{ID: "partner", Name: "Partner", JoinedDate: "2026-04-14", Currency: "USD"}},
			},
		},
	}, true, accountContentLocale(settings))

	if message.HasPayload || len(message.Items) != 0 {
		t.Fatalf("expected one-time buyout collection reminder to be skipped, got %#v", message.Items)
	}
}

func TestNotificationScheduleCandidateSubscriptionsMatchFullFiltering(t *testing.T) {
	app := newSchemaTestApp(t)
	if err := ensureSchema(app); err != nil {
		t.Fatal(err)
	}
	user, _ := createRouteTestUser(t, app, "notification-candidates")
	settings := defaultAppSettings()
	settings.Timezone = "UTC"
	settings.NotificationTimeLocal = "08:00"
	settings.NotificationReminderDays = 5
	settings.ShowExpired = true
	schedule := localScheduleOccurrence{
		ScheduledLocalDate:  "2026-05-14",
		ScheduledLocalTime:  "08:00",
		TimeZone:            "UTC",
		ScheduledInstantUTC: "2026-05-14T08:00:00Z",
	}

	createRouteTestSubscription(t, app, user.Id, map[string]interface{}{"name": "Renewal", "nextBillingDate": "2026-05-17", "reminderDays": 3})
	createRouteTestSubscription(t, app, user.Id, map[string]interface{}{"name": "Inherited", "nextBillingDate": "2026-05-19", "reminderDays": inheritReminderDays})
	createRouteTestSubscription(t, app, user.Id, map[string]interface{}{"name": "Trial", "status": "trial", "nextBillingDate": "2026-06-01", "trialEndDate": "2026-05-15", "reminderDays": 1})
	createRouteTestSubscription(t, app, user.Id, map[string]interface{}{"name": "Expired", "nextBillingDate": "2026-05-01", "reminderDays": 7})
	createRouteTestSubscription(t, app, user.Id, map[string]interface{}{"name": "Fixed Term", "billingCycle": "one-time", "oneTimeTermCount": 6, "oneTimeTermUnit": "month", "nextBillingDate": "2026-05-17", "reminderDays": 3})
	createRouteTestSubscription(t, app, user.Id, map[string]interface{}{"name": "Lifetime", "billingCycle": "one-time", "nextBillingDate": "2026-05-14", "reminderDays": 0})
	createRouteTestSubscription(t, app, user.Id, map[string]interface{}{"name": "Quiet", "nextBillingDate": "2026-05-17", "reminderDays": disabledReminderDays})
	familyCostSharing := types.JSONRaw(`{"enabled":true,"splitMode":"equal","collectionReminder":{"enabled":true,"reminderDays":3},"members":[{"id":"partner","name":"Partner","joinedDate":"2026-04-17","currency":"USD"}]}`)
	createRouteTestSubscription(t, app, user.Id, map[string]interface{}{"name": "Family Collection", "price": "30", "nextBillingDate": "2026-05-17", "reminderDays": disabledReminderDays, "costSharingCollectionReminderEnabled": true, "costSharingNextCollectionReminderDate": "2026-05-14", "costSharing": familyCostSharing})
	createRouteTestSubscription(t, app, user.Id, map[string]interface{}{"name": "Future", "nextBillingDate": "2040-01-01", "reminderDays": 3})

	full, err := listNotificationSubscriptions(app, user.Id)
	if err != nil {
		t.Fatal(err)
	}
	candidates, err := listNotificationScheduleCandidateSubscriptions(app, user.Id, settings, schedule, true)
	if err != nil {
		t.Fatal(err)
	}
	fullMessage := buildDueNotificationForSchedule(schedule, time.Date(2026, 5, 14, 8, 0, 0, 0, time.UTC), settings, full, true, accountContentLocale(settings))
	candidateMessage := buildDueNotificationForSchedule(schedule, time.Date(2026, 5, 14, 8, 0, 0, 0, time.UTC), settings, candidates, true, accountContentLocale(settings))

	if len(candidates) >= len(full) {
		t.Fatalf("expected cron candidates to avoid full subscription scan, got candidates=%d full=%d", len(candidates), len(full))
	}
	if got, want := notificationItemKeys(candidateMessage.Items), notificationItemKeys(fullMessage.Items); strings.Join(got, "|") != strings.Join(want, "|") {
		t.Fatalf("candidate items = %#v, want %#v", got, want)
	}
}

func TestNotificationCronSettlesExhaustedCostSharingFailedJob(t *testing.T) {
	for _, tc := range []struct {
		name       string
		attempts   int
		maxRetries string
		wantReason string
	}{
		{name: "max retries reached", attempts: 3, wantReason: "max_retries_reached"},
		{name: "retries disabled", attempts: 1, maxRetries: "0", wantReason: "retries_disabled"},
	} {
		t.Run(tc.name, func(t *testing.T) {
			if tc.maxRetries != "" {
				t.Setenv("NOTIFICATION_MAX_RETRIES", tc.maxRetries)
			}
			app := newSchemaTestApp(t)
			if err := ensureSchema(app); err != nil {
				t.Fatal(err)
			}
			user, _ := createRouteTestUser(t, app, "cost-sharing-exhausted-"+strings.ReplaceAll(tc.name, " ", "-"))
			settings := defaultAppSettings()
			settings.Timezone = "UTC"
			settings.NotificationTimeLocal = "08:00"
			settings.EnabledChannels = []string{"webhook"}
			createNotificationCronRouteTestSettings(t, app, user, settings)
			costSharing := types.JSONRaw(`{"enabled":true,"splitMode":"equal","collectionReminder":{"enabled":true,"reminderDays":3},"members":[{"id":"partner","name":"Partner","joinedDate":"2026-04-17","currency":"USD"}]}`)
			subscription := createRouteTestSubscription(t, app, user.Id, map[string]interface{}{
				"name":                                  "Family Collection Exhausted",
				"price":                                 "30",
				"nextBillingDate":                       "2026-05-17",
				"reminderDays":                          disabledReminderDays,
				"costSharing":                           costSharing,
				"costSharingCollectionReminderEnabled":  true,
				"costSharingNextCollectionReminderDate": "2026-05-14",
			})
			now := time.Date(2026, 5, 14, 8, 0, 0, 0, time.UTC)
			refreshNotificationSchedulerForTest(t, app, user.Id, now)
			createFailedCronJobForTest(t, app, user.Id, settings, now, tc.attempts)

			sendCount := 0
			originalSender := notificationSenders["webhook"]
			notificationSenders["webhook"] = notificationSenderFunc(func(_ core.App, _ appSettings, _ notificationMessage, _ appLocale) error {
				sendCount++
				return errors.New("still failing")
			})
			t.Cleanup(func() {
				notificationSenders["webhook"] = originalSender
			})

			result, err := runNotificationCron(app, notificationCronOptions{Now: now, WindowMinutes: 2})
			if err != nil {
				t.Fatal(err)
			}
			if result.Skipped != 1 || len(result.Results) != 1 || result.Results[0].Reason != tc.wantReason {
				t.Fatalf("expected exhausted failed job to settle as %s, got %#v", tc.wantReason, result)
			}
			if sendCount != 0 {
				t.Fatalf("expected exhausted failed job not to send again, sent %d times", sendCount)
			}
			reloaded, err := app.FindRecordById("subscriptions", subscription.Id)
			if err != nil {
				t.Fatal(err)
			}
			if got := reloaded.GetString("costSharingNextCollectionReminderDate"); got != "2026-06-14" {
				t.Fatalf("expected collection reminder mirror to advance, got %q", got)
			}
			job, err := getNotificationJob(app, user.Id, "2026-05-14", "08:00", "UTC")
			if err != nil {
				t.Fatal(err)
			}
			if job.GetString("status") != notificationStatusFailed || job.GetInt("attempts") != tc.attempts {
				t.Fatalf("expected failed job history to remain unchanged, status=%q attempts=%d", job.GetString("status"), job.GetInt("attempts"))
			}
		})
	}
}

func TestUpcomingBatchKeepsSameDayCostSharingMembers(t *testing.T) {
	batches := map[string]*upcomingNotificationBatch{}
	occurrence := localScheduleOccurrence{
		ScheduledLocalDate:  "2026-05-14",
		ScheduledLocalTime:  "08:00",
		TimeZone:            "UTC",
		ScheduledInstantUTC: "2026-05-14T08:00:00Z",
	}
	sub := notificationSubscription{
		ID:              "sub_family",
		Name:            "Family Plan",
		Price:           "30",
		Currency:        "USD",
		Status:          "active",
		BillingCycle:    "monthly",
		NextBillingDate: "2026-05-17",
		ReminderDays:    3,
	}

	appendUpcomingBatch(batches, occurrence, []notificationContentItem{
		newCostSharingNotificationContentItem(sub, "Partner", "10", "USD", "2026-05-17", 3, 3),
		newCostSharingNotificationContentItem(sub, "Child", "10", "USD", "2026-05-17", 3, 3),
	})

	batch := batches["2026-05-14|08:00|UTC"]
	if batch == nil || len(batch.Items) != 2 {
		t.Fatalf("expected two member collection reminders, got %#v", batch)
	}
	if got := notificationItemKeys(batch.Items); strings.Join(got, "|") != "costSharing|Family Plan|2026-05-17||Child/10/USD|costSharing|Family Plan|2026-05-17||Partner/10/USD" {
		t.Fatalf("unexpected collection reminder keys: %#v", got)
	}
}

func createFailedCronJobForTest(t *testing.T, app core.App, userID string, settings appSettings, now time.Time, attempts int) {
	t.Helper()
	schedule := getLocalScheduleDecision(now, settings.Timezone, settings.NotificationTimeLocal, 2, false)
	job, _, err := createNotificationJob(app, userID, schedule, notificationStatusFailed, attempts)
	if err != nil {
		t.Fatal(err)
	}
	result := createJobResult(
		"some_channels_failed",
		schedule.localScheduleOccurrence,
		settings,
		accountContentLocale(settings),
		notificationMessage{
			Title:      "Renewlet",
			Content:    "failed",
			Timestamp:  "2026-05-14 08:00:00 UTC",
			HasPayload: true,
		},
		notificationCronOptions{Now: now, WindowMinutes: 2},
		jobChannels{
			Attempted: []string{"webhook"},
			Failed:    []channelFailure{{Channel: "webhook", Error: "still failing"}},
		},
	)
	if err := finalizeNotificationJob(app, job, userID, schedule, notificationStatusFailed, "webhook: still failing", result); err != nil {
		t.Fatal(err)
	}
}
